pg_search/src/postgres/merge.rs

```diff
@@ -0,0 +1,286 @@
+        self.foreground_layer_sizes.clone()
+    }
+
+    fn background(&self) -> Vec<u64> {
+        self.background_layer_sizes.clone()
+    }
+}
+
+/// Kick off a merge of the index, if needed.
+///
+/// First merge into the smaller layers in the foreground,
+/// then launch a background worker to merge down the larger layers.
+pub unsafe fn do_merge(index: &PgSearchRelation) -> anyhow::Result<()> {
+    let merger = SearchIndexMerger::open(MvccSatisfies::Mergeable.directory(index))?;
+    let layer_sizes = IndexLayerSizes::from(index);
+    let foreground_layers = layer_sizes.foreground();
+    let background_layers = layer_sizes.background();
+
+    let metadata = MetaPage::open(index);
+    let merge_lock = metadata.acquire_merge_lock();
+
+    let needs_background_merge = !background_layers.is_empty() && {
+        let mut background_merge_policy = LayeredMergePolicy::new(background_layers.clone());
+        background_merge_policy.set_mergeable_segment_entries(&metadata, &merge_lock, &merger);
+        let merge_candidates = background_merge_policy.simulate();
+        !merge_candidates.is_empty()
+    };
+
+    // first merge down the foreground layers
+    if !foreground_layers.is_empty() {
+        let foreground_merge_policy = LayeredMergePolicy::new(foreground_layers);
+        unsafe { foreground_merge_policy.merge_index(index, merge_lock, false) };
+    }
+
+    pgrx::debug1!("foreground merge complete, needs_background_merge: {needs_background_merge}");
+
+    // then launch a background process to merge down the background layers
+    // only if we determine that there are enough segments to merge in the background
+    if needs_background_merge {
+        try_launch_background_merger(index);
+    }
+
+    Ok(())
+}
+
+/// Try to launch a background process to merge down the index.
+/// Is not guaranteed to launch the process if there are not enough `max_worker_processes` available.
+unsafe fn try_launch_background_merger(index: &PgSearchRelation) {
+    let dbname = CStr::from_ptr(pg_sys::get_database_name(pg_sys::MyDatabaseId))
+        .to_string_lossy()
+        .into_owned();
+
+    let worker_name = format!(
+        "background merger for {}.{}",
+        index.namespace(),
+        index.name()
+    );
+    if BackgroundWorkerBuilder::new(&worker_name)
+        .enable_spi_access()
+        .enable_shmem_access(None)
+        .set_library("pg_search")
+        .set_function("background_merge")
+        .set_argument(index.oid().into_datum())
+        .set_extra(&dbname)
+        .set_notify_pid(unsafe { pg_sys::MyProcPid })
+        .load_dynamic()
+        .is_err()
+    {
+        ErrorReport::new(
+            PgSqlErrorCode::ERRCODE_INSUFFICIENT_RESOURCES,
+            "not enough available `max_worker_processes` to launch a background merger",
+            function_name!(),
+        )
+        .set_hint("`SET max_worker_processes = <number>`")
+        .report(PgLogLevel::WARNING);
```

Comment:
Perhaps contrary to what @stuhood said above, I don't think this even warrants a WARNING message.

We could `PgLogLevel::LOG` it, but WARNINGs tend to find their way back to the client and definitely show up interactively via psql, and there's nothing the user can do about this, per se.

Part of the reason I don't think it's necessary is, I assume, if the worker did start it could then determine that there's actually no work for it to do, and that's really not a lot different as far as the user is concerned.

tl;dr:  Make this `PgLogLevel::LOG`.  This way it's always emitted to the log but not in the user's face.